aspnetcore/blazor/components/css-isolation.md

```diff
@@ -0,0 +1,148 @@
+author: daveabrock
+description: Learn how CSS isolation allows you to scope CSS to your components, which can simplify your CSS and avoid collisions with other components or libraries.
+monikerRange: '>= aspnetcore-5.0'
+ms.author: riande
+ms.custom: mvc
+ms.date: 09/17/2020
+no-loc: ["ASP.NET Core Identity", cookie, Cookie, Blazor, "Blazor Server", "Blazor WebAssembly", "Identity", "Let's Encrypt", Razor, SignalR]
+uid: blazor/components/css-isolation
+---
+# ASP.NET Core Blazor CSS isolation
+
+By [Dave Brock](https://twitter.com/daveabrock)
+
+Use cascading style sheets (CSS) isolation to only apply styles to the current component. CSS isolation simplifies an app's CSS footprint by preventing dependencies on global styles, and also helps to avoid styling conflicts between other components or libraries.
+
+## Enable CSS isolation 
+
+To define component-specific styles, create a `razor.css` file matching the name of the `.razor` file for the component. This `razor.css` file is a *scoped CSS file*. 
+
+For a `MyComponent` that has a `MyComponent.razor` file, create a file in the `Pages` folder called `MyComponent.razor.css`. The `MyComponent` value in the `razor.css` filename is **not** case-sensitive.
+
+For example, to add CSS isolation to the `Counter` component in the sample app, add a new file named `Counter.razor.css`, then add the following CSS:
+
+```css
+h1 { 
+    color: brown;
+    font-family: Tahoma, Geneva, Verdana, sans-serif;
+}
+```
+
+The styles defined in `Counter.razor.css` are only applied to the rendered output of the `Counter` component. Any `h1` CSS declarations defined elsewhere do not conflict with `Counter` styles.
+
+In order to guarantee style isolation when bundling occurs, `@import` blocks are not supported with scoped CSS files.
+
+## Child component support
+
+By default, CSS isolation only applies to the component you associate with `MyComponent.razor.css`. To apply changes to a child component, use the `::deep` combinator in the parent component's `razor.css` file.
+
+The following example shows a parent component containing a child component, `CounterChild`, in `Pages/Counter.razor`:
+
+```razor
+@page "/counter"
+
+<h1>Counter</h1>
+
+<div>
+    <p>Current count: @currentCount</p>
+
+    <button class="btn btn-primary" @onclick="IncrementCount">Click me</button>
+
+    <CounterChild />
+</div>
+```
+
+The `CounterChild` component contains an `h1` element, to be styled the same as `Counter`:
+
+```razor
+@page "/counter-child"
+
+<h1>Child component</h1>
+```
+
+Update the `h1` declaration in `Counter.razor.css` with the `::deep` combinator to signify the `h1` style declaration must cascade to the child component.
+
+```css
+::deep h1 { 
+    color: brown;
+    font-family: Tahoma, Geneva, Verdana, sans-serif;
+}
+
+```
+
+The `h1` style now applies to both the `Counter` and `CounterChild` components without the need to create a separate scoped CSS file for `CounterChild`.
```

Comment:
In this case the selector applies to the Counter and CounterChild components because the `h1`s are not at the top level, but are wrapped into a `div`. `::deep` only selects elements that are descendants of an element with the generated attribute (`[b-...]`).

````razor
@page "/counter"
<h1>Not styled by scoped CSS with ::deep</h1>
<div>
    <h1>Styled by scoped CSS</h1>
</div>
````

````html
<!-- [b-a92ooaz332] h1 doesn't match this element -->
<h1 b-a92ooaz332>Not styled by scoped CSS with ::deep</h1> 
<div b-a92ooaz332>
    <h1 b-a92ooaz332>Styled by scoped CSS</h1>
</div>
````

`h1, ::deep h1` would select any `h1` in both components. I think this behavior should be documented somewhere.

Reply:
I actually found that if you do this, the parent component gets styled `red` and the children `brown`...

```css
::deep h1 {
    color: red;
}

h1 {
    color: brown;
}
```

...and precedence matters! That is, if you do `h1` first and `::deep h1` second, the parent and child selector is both red.

Then to make them all red, `::deep h1` works without the need to do `h1, ::deep h1`. I would appreciate if you gave this a shot to prove I'm not (too) crazy. 😀

Reply:
The declaration order should not matter here. `::deep h1` is translated to `[b-xxx] h1`. This selector is more specific ([specificity](https://developer.mozilla.org/en-US/docs/Web/CSS/Specificity)) than `h1`, so `::deep h1` overrides `h1` when both selectors apply to an element.

Using the Counter and CounterChild components, the h1 is brown in the parent component, and red in the child component. It's the same if you change the declaration order.

Let's have fun with the `::deep` pseudo-selector 😊

For all examples 
- the style is `::deep h1 { color: red; }`
- the child component is `<h1>Child component</h1>`

### Example 1

````
<h1>Counter</h1>
<div>
    <CounterChild />
</div>
````

Parent h1: unstyled. Child h1: red.

### Example 2

````
<h1>Counter</h1>
<CounterChild />
````

Parent h1: unstyled. Child h1: unstyled.

### Example 3

````
<div>
    <h1>Counter</h1>
    <CounterChild />
</div>
````

Parent h1: red. Child h1: red.


### Example 4

````
<div>
    <h1>Counter</h1>
</div>
<CounterChild />
````

Parent h1: red. Child h1: unstyled.

> "To apply changes to a child component, use the `::deep` combinator in the parent component's `razor.css` file."

This doesn't seem to be true for example 2 or 4. If you can validate my examples, I think it would make sense to add a note that a `div` is required for the `::deep` selector to work as expected.

Reply:
I've edited the first part of my previous comment because I misread your comment :(

Reply:
Thanks for the great clarificationsI I'll make my updates tonight after the day job and push to the PR.

Reply:
Updated the docs with my latest push. Thanks again for the great feedback!

Reply:
The latest version is very clear! Good job :)